From 5e4b95f92cd93df5a1a83b277b669c43329b1d48 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:09:46 -0400
Subject: [virt] kvm/ppc: Set the runlatch bit of a CPU just before starting guest

Message-id: <1410545655-205645-358-git-send-email-dzickus@redhat.com>
Patchwork-id: 94243
O-Subject: [RHEL7 PATCH 357/626] ppc/kvm: Set the runlatch bit of a CPU just before starting guest
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit fd17dc7b9a8d57150ee0121c4c83edbb7e26eece
Author: Preeti U Murthy <preeti@linux.vnet.ibm.com>
Date:   Fri Apr 11 16:01:58 2014 +0530

    ppc/kvm: Set the runlatch bit of a CPU just before starting guest

    The secondary threads in the core are kept offline before launching guests
    in kvm on powerpc: "371fefd6f2dc4666:KVM: PPC: Allow book3s_hv guests to use
    SMT processor modes."

    Hence their runlatch bits are cleared. When the secondary threads are called
    in to start a guest, their runlatch bits need to be set to indicate that they
    are busy. The primary thread has its runlatch bit set though, but there is no
    harm in setting this bit once again. Hence set the runlatch bit for all
    threads before they start guest.

    Signed-off-by: Preeti U Murthy <preeti@linux.vnet.ibm.com>
    Acked-by: Paul Mackerras <paulus@samba.org>
    Reviewed-by: Srivatsa S. Bhat <srivatsa.bhat@linux.vnet.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kvm/book3s_hv_rmhandlers.S b/arch/powerpc/kvm/book3s_hv_rmhandlers.S
index 96236ff..be2e8f6 100644
--- a/arch/powerpc/kvm/book3s_hv_rmhandlers.S
+++ b/arch/powerpc/kvm/book3s_hv_rmhandlers.S
@@ -240,6 +240,12 @@ kvm_novcpu_exit:
  */
  .globl kvm_start_guest
 kvm_start_guest:
+
+ /* Set runlatch bit the minute you wake up from nap */
+ mfspr r1, SPRN_CTRLF
+ ori  r1, r1, 1
+ mtspr SPRN_CTRLT, r1
+
  ld r2,PACATOC(r13)
 
  li r0,KVM_HWTHREAD_IN_KVM
-- 
1.7.1